valohai_cli/utils/ssh.py

```diff
@@ -0,0 +1,131 @@
+        click.echo(f"Failed to execute SSH command: {e}", err=True)
+        if e.stderr:
+            click.echo(f"Error details: {e.stderr}", err=True)
+
+
+def find_ssh_details(events: list) -> Optional[Tuple[str, int]]:
+    prefix = "::ssh::"
+    for event in events:
+        if event["message"].startswith(prefix):
+            data = json.loads(event["message"][len(prefix) :])
+            return data["address"], int(data["port"])
+    return None
+
+
+def get_default_ssh_directory() -> str:
+    return os.path.expanduser("~/.ssh")
+
+
+def get_ssh_key_pairs_from_dir(dir: str) -> Iterator[Keypair]:
+    folder_path = pathlib.Path(dir)
+    for pth in folder_path.glob("*.pub"):
+        if not pth.is_file():
+            continue
+        try:
+            with pth.open("rb") as f:
+                first_line = f.readline().strip()
+                if first_line.startswith((b"ssh-rsa", b"ecdsa-sha2-", b"ssh-ed25519")):
+                    private_key_paths = [pth.with_suffix(""), pth.with_suffix(".pem")]
+                    for private_key_path in private_key_paths:
+                        if private_key_path.is_file():
+                            yield Keypair(public_path=pth, private_path=private_key_path)
+                            break
+        except OSError:  # e.g. unreadable key
+            pass
+
+
+def select_private_key_from_possible_directories() -> Optional[str]:
+    default_ssh_dir = get_default_ssh_directory()
+    cli_config_dir = get_settings_file_name("")
+    all_keypairs: list[Keypair] = []
+    for dir in [cli_config_dir, default_ssh_dir]:
+        click.echo(f"Looking for private key in {dir}...")
+        if not os.path.exists(dir):
+            continue
+        if key_pairs := get_ssh_key_pairs_from_dir(dir):
+            all_keypairs.extend(key_pairs)
+
+    if not all_keypairs:
+        return None
+
+    keypair_options = [
+        {
+            "name": keypair.private_path,
+            "description": f"Public: {keypair.public_path}",
+        }
+        for keypair in all_keypairs
+    ]
+
+    selected_key = prompt_from_list(prompt="Select the SSH key pair you want to use", options=keypair_options)
+    return str(selected_key["name"])
+
+
+def fetch_status_events(execution: dict, first_n: Optional[int] = None) -> Generator:
+    params = {}
+    if first_n is not None:
+        params["first_n"] = first_n
+    events_response = request("get", f"{execution['url']}status_events/", params=params).json()
+    yield from events_response.get("status_events", [])
+
+
+def get_ssh_details_from_execution(counter: int) -> tuple[str, int]:
+    project = get_project(require=True)
+    assert project
+    execution = project.get_execution_from_counter(
+        counter=counter,
+        params={
+            "exclude": "metadata, events",
+        },
+    )
+    events_response = fetch_status_events(execution, first_n=100)
+    ssh_details = find_ssh_details(list(events_response))
```

Comment:
```suggestion
    ssh_details = find_ssh_details(fetch_status_events(execution, first_n=100))
```
(and make the argument type accepted by `find_ssh_details` an `Iterable`)